clang/test/CIR/CodeGen/X86/avx512dq-builtins.c

```diff
@@ -0,0 +1,213 @@
+
+  // LLVM-LABEL: _kandn_mask8
+  // LLVM: [[L:%.*]] = bitcast i8 %{{.*}} to <8 x i1>
+  // LLVM: [[R:%.*]] = bitcast i8 %{{.*}} to <8 x i1>
+  // LLVM: xor <8 x i1> [[L]], {{.*}}
+  // LLVM: and <8 x i1>
+  // LLVM: bitcast <8 x i1> {{.*}} to i8
+
+  // OGCG-LABEL: _kandn_mask8
+  // OGCG: bitcast i8 %{{.*}} to <8 x i1>
+  // OGCG: bitcast i8 %{{.*}} to <8 x i1>
+  // OGCG: xor <8 x i1>
+  // OGCG: and <8 x i1>
+  // OGCG: bitcast <8 x i1> {{.*}} to i8
+
+  return _kandn_mask8(A, B);
+}
+
+__mmask8 test_kor_mask8(__mmask8 A, __mmask8 B) {
+  // CIR-LABEL: _kor_mask8
+  // CIR: cir.cast bitcast {{.*}} : !u8i -> !cir.vector<8 x !cir.int<s, 1>>
+  // CIR: cir.cast bitcast {{.*}} : !u8i -> !cir.vector<8 x !cir.int<s, 1>>
+  // CIR: cir.binop(or, {{.*}}, {{.*}}) : !cir.vector<8 x !cir.int<s, 1>>
+  // CIR: cir.cast bitcast {{.*}} : !cir.vector<8 x !cir.int<s, 1>> -> !u8i
+
+  // LLVM-LABEL: _kor_mask8
+  // LLVM: [[L:%.*]] = bitcast i8 %{{.*}} to <8 x i1>
+  // LLVM: [[R:%.*]] = bitcast i8 %{{.*}} to <8 x i1>
+  // LLVM: or <8 x i1> [[L]], [[R]]
+  // LLVM: bitcast <8 x i1> {{.*}} to i8
+
+  // OGCG-LABEL: _kor_mask8
+  // OGCG: bitcast i8 %{{.*}} to <8 x i1>
+  // OGCG: bitcast i8 %{{.*}} to <8 x i1>
+  // OGCG: or <8 x i1>
+  // OGCG: bitcast <8 x i1> {{.*}} to i8
+
+  return _kor_mask8(A, B);
+}
+
+__mmask8 test_kxor_mask8(__mmask8 A, __mmask8 B) {
+  // CIR-LABEL: _kxor_mask8
+  // CIR: cir.cast bitcast {{.*}} : !u8i -> !cir.vector<8 x !cir.int<s, 1>>
+  // CIR: cir.cast bitcast {{.*}} : !u8i -> !cir.vector<8 x !cir.int<s, 1>>
+  // CIR: cir.binop(xor, {{.*}}, {{.*}}) : !cir.vector<8 x !cir.int<s, 1>>
+  // CIR: cir.cast bitcast {{.*}} : !cir.vector<8 x !cir.int<s, 1>> -> !u8i
+
+  // LLVM-LABEL: _kxor_mask8
+  // LLVM: [[L:%.*]] = bitcast i8 %{{.*}} to <8 x i1>
+  // LLVM: [[R:%.*]] = bitcast i8 %{{.*}} to <8 x i1>
+  // LLVM: xor <8 x i1> [[L]], [[R]]
+  // LLVM: bitcast <8 x i1> {{.*}} to i8
+
+  // OGCG-LABEL: _kxor_mask8
+  // OGCG: bitcast i8 %{{.*}} to <8 x i1>
+  // OGCG: bitcast i8 %{{.*}} to <8 x i1>
+  // OGCG: xor <8 x i1>
+  // OGCG: bitcast <8 x i1> {{.*}} to i8
+
+  return _kxor_mask8(A, B);
+}
+
+__mmask8 test_kxnor_mask8(__mmask8 A, __mmask8 B) {
+  // CIR-LABEL: _kxnor_mask8
+  // CIR: cir.cast bitcast {{.*}} : !u8i -> !cir.vector<8 x !cir.int<s, 1>>
+  // CIR: cir.cast bitcast {{.*}} : !u8i -> !cir.vector<8 x !cir.int<s, 1>>
+  // CIR: cir.unary(not, {{.*}}) : !cir.vector<8 x !cir.int<s, 1>>
+  // CIR: cir.binop(xor, {{.*}}, {{.*}}) : !cir.vector<8 x !cir.int<s, 1>>
+  // CIR: cir.cast bitcast {{.*}} : !cir.vector<8 x !cir.int<s, 1>> -> !u8i
+
+  // LLVM-LABEL: _kxnor_mask8
+  // LLVM: [[L:%.*]] = bitcast i8 %{{.*}} to <8 x i1>
+  // LLVM: [[R:%.*]] = bitcast i8 %{{.*}} to <8 x i1>
+  // LLVM: xor <8 x i1> [[L]], {{.*}}
+  // LLVM: xor <8 x i1> {{.*}}, [[R]]
+
+  // OGCG-LABEL: _kxnor_mask8
+  // OGCG: bitcast i8 %{{.*}} to <8 x i1>
+  // OGCG: bitcast i8 %{{.*}} to <8 x i1>
+  // OGCG: xor <8 x i1>

```

Comment:
Perhaps also check for `splat (i1 true)` for the first xor (here and in other places where the xor instruction is used to invert the value)